enhancements/ingress/optional-ingress-hypershift.md

```diff
@@ -0,0 +1,272 @@
+The required capabilities should be applied to [the cluster version operator manifests](https://github.com/openshift/cluster-version-operator/blob/master/install/0000_00_cluster-version-operator_03_deployment.yaml) only.
+This should effectively descope them from the HyperShift installations since HyperShift manages [the cluster version operator's deployment directly](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318).
+
+#### Ingress operator APIs
+
+The following APIs are not available if the ingress capability is disabled:
+- CRDs which are part of the cluster operator's manifest payload:
+  - [IngressController](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition.yaml)
+  - [DNSRecord](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition-internal.yaml)
+- CRDs which are created at runtime if `gateway-api` feature gate is set:
+  - [GatewayClass](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gatewayclasses.yaml)
+  - [Gateway](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gateways.yaml)
+  - [HTTPRoute](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_httproutes.yaml)
+
+#### Related APIs
+
+The following APIs are left intact if the ingress capability is disabled:
+- [Route API](https://docs.openshift.com/container-platform/4.13/rest_api/network_apis/route-route-openshift-io-v1.html). Implemented in two places: OpenShift API as [CRD](https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml) and in [openshift-api-server](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).
+- [Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html). Implemented in OpenShift API as [CRD](https://github.com/openshift/api/blob/master/config/v1/0000_10_config-operator_01_ingress.crd.yaml).
+
+However the implementation of these APIs (or parts of them) depend on the ingress operator:
+- The ingress operator creates the default ingress controller to manage the backends for the routes.
+- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role bindings to get the custom TLS certificate secret from `openshift-config` namespace.
+
+#### Console operator
+
+To prevent the console operator from going degraded, [the console route check](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/healthcheck/controller.go#L122) should be skipped if the ingress capability is not enabled in `ClusterVersion`.
```

Comment:
So this is not the only usage for which we need the ingress config.
Console-operator also needs the ingress config for:
- [CLIDownloads controlle](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/clidownloads/controller.go#L118) where we use it to assemble the URL for the CLI tools
- [Routes controller](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/route/controller.go#L121) where we use it to assemble the URL for all console related route's 
- [Service controller](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/service/controller.go#L109) 

To sum it up, ingress config is used for all the logic related to console routes. Due to that reason I dont think we skip can just skip this logic, until we have other config which could substitute the ingress one.

Other option would be to disable the console capability if ingress capability is disabled.

Reply:
@jhadvig If the config is present, can we just ignore that routes are not admitted and/or available as endpoints if ingress is not enabled?

Reply:
what is all of this going to mean in terms of the console experience on a hypershift cluster that has ingress disabled?  Are we expecting the console to create LB ingress instead?

Reply:
It's the same problem as for any component which depends on the Route API. The long term solution is something we will need to discuss, one of the idea is described in the [Add Route API to ingress capability alternative section](https://github.com/openshift/enhancements/blob/e8779ca74822d9b425ed48e264414c2a5a8b038a/enhancements/ingress/optional-ingress-hypershift.md#add-route-api-to-the-ingress-capability).  I'm not sure the answer is known by anybody for the moment. The request is coming from ROSA, the cluster ingress operator and HyperShift are just stepping stones. The decision is likely to be taken by ROSA engineers. This EP tries to address the requirements of the RFE-3395 which [leaves this decision to the future](https://issues.redhat.com/browse/RFE-3395?focusedId=21514823&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-21514823). Also the RFE states (in [the same thread](https://issues.redhat.com/browse/RFE-3395?focusedId=21514823&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-21514823)) that AWS Load Balancer Operator will be available, so a Kubernetes service of type Load Balancer or an Ingress can do the thing eventually.

Reply:
> It's the same problem as for any component which depends on the Route API. The long term solution is something we will need to discuss

that's fine, but this EP needs to be very clear about what the behavior/UX will be in the short term when this is disabled.  I don't think it does so yet.

Reply:
@jhadvig can you respond to https://github.com/openshift/enhancements/pull/1415/files#r1228265620 :

> @jhadvig If the config is present, can we just ignore that routes are not admitted and/or available as endpoints if ingress is not enabled?

this seems like one of the remaining open issues on this EP before it can be approved (ensuring we understand the full implications on Console of disabling ingress, and what mitigation actions will be taken)

Reply:
Sorry, my bad. We discussed this with @jhadvig and reached the agreement that the console will stay running but without the ingress served by OpneShift ingress controller. The idea was similar to the one already discussed in the EP: the consumer of HyperShift (ROSA) will have to use another means to accept the ingress traffic for the console (AWS LB Operator is the candidate).